google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClientImpl.java

```diff
@@ -383,27 +418,58 @@ private Future<Dialect> getDialectAsync() {
     return pool.getDialectAsync();
   }
 
+  private UpdateOption[] withReqId(
+      final XGoogSpannerRequestId reqId, final UpdateOption... options) {
+    if (reqId == null) {
+      return options;
+    }
+    ArrayList<UpdateOption> allOptions = new ArrayList(Arrays.asList(options));
+    allOptions.add(new Options.RequestIdOption(reqId));
+    return allOptions.toArray(new UpdateOption[0]);
+  }
+
+  private TransactionOption[] withReqId(
+      final XGoogSpannerRequestId reqId, final TransactionOption... options) {
+    if (reqId == null) {
+      return options;
+    }
+    ArrayList<TransactionOption> allOptions = new ArrayList(Arrays.asList(options));
+    allOptions.add(new Options.RequestIdOption(reqId));
+    return allOptions.toArray(new TransactionOption[0]);
+  }
+
   private long executePartitionedUpdateWithPooledSession(
       final Statement stmt, final UpdateOption... options) {
     ISpan span = tracer.spanBuilder(PARTITION_DML_TRANSACTION, commonAttributes);
     try (IScope s = tracer.withSpan(span)) {
-      return runWithSessionRetry(session -> session.executePartitionedUpdate(stmt, options));
+      return runWithSessionRetry(
+          (session, reqId) -> {
+            return session.executePartitionedUpdate(stmt, withReqId(reqId, options));
+          });
     } catch (RuntimeException e) {
       span.setStatus(e);
       span.end();
       throw e;
     }
   }
 
-  private <T> T runWithSessionRetry(Function<Session, T> callable) {
+  @VisibleForTesting
+  <T> T runWithSessionRetry(BiFunction<Session, XGoogSpannerRequestId, T> callable) {
     PooledSessionFuture session = getSession();
+    XGoogSpannerRequestId reqId =
+        XGoogSpannerRequestId.of(
+            this.dbId, Long.valueOf(session.getChannel()), this.nextNthRequest(), 0);
     while (true) {
       try {
-        return callable.apply(session);
+        reqId.incrementAttempt();
```

Comment:
This type of retry should not cause the attempt number to be increased. This retry function is for the (relatively uncommon) case that Spanner returns `SessionNotFound`. This will generate a new request, using a new session etc. and should not be treated as a transparent retry.

Reply:
Roger that and thank you, I've made the updates!